src/execution/lut.rs

```diff
@@ -0,0 +1,113 @@
+            // operation. Should it be a RuntimeError if said missing pair is called?
+
+            function_lut.push(module_lut);
+        }
+
+        Some(Self { function_lut })
+    }
+
+    /// Lookup a function by its module and function index.
+    ///
+    /// # Arguments
+    /// - `module_idx`: The index of the module within the `modules` array passed in [Lut::new].
+    /// - `function_idx`: The index of the function within the module. This index is considered in-bounds only if it is
+    ///   an index of an imported function.
+    ///
+    /// # Returns
+    /// - `None`, if the indicies are out of bound
+    /// - `Some(export_module_idx, export_function_idx)`, where the new indicies are the indicies of the module which
+    ///   contains the implementation of the imported function, and the implementation has the returned index within.
+    pub fn lookup(&self, module_idx: usize, function_idx: usize) -> Option<(usize, usize)> {
+        self.function_lut
+            .get(module_idx)?
+            .get(function_idx)
+            .copied()
+    }
+
+    /// Manually lookup a function by its module and function name.
+    ///
+    /// This function is used to resolve import directives before the [Lut] is created, and can be used to resolve
+    /// imports even after the [Lut] is created at the cost of speed.
+    ///
+    /// # Arguments
+    /// - `modules`: The modules to link together.
+    /// - `module_map`: A map from module name to module index within the `modules` array.
+    /// - `module_name`: The name of the module which imports the function.
+    /// - `function_name`: The name of the function to import.
+    ///
+    /// # Returns
+    /// - `None`, if the module or function is not found.
+    /// - `Some(export_module_idx, export_function_idx)`, where the new indicies are the indicies of the module which
+    ///    contains the implementation of the imported function, and the implementation has the returned index within.
+    ///    Note that this function returns the first matching function, if there are multiple functions with the same
+    ///    name.
+    pub fn manual_lookup(
+        modules: &[ExecutionInfo],
+        module_map: &BTreeMap<String, usize>,
+        module_name: &str,
+        function_name: &str,
+    ) -> Option<(usize, usize)> {
+        let module_idx = module_map.get(module_name)?;
+        let module = &modules[*module_idx];
+
+        module
+            .store
+            .exports
+            .iter()
+            .filter_map(|export| {
+                if export.name == function_name {
+                    Some(&export.desc)
+                } else {
+                    None
+                }
+            })
+            .find_map(|desc| {
+                if let ExportDesc::FuncIdx(func_idx) = desc {
+                    Some((*module_idx, *func_idx))
+                } else {
+                    None
+                }
+            })
+    }
+}

```

Comment:
This scheme can't handle cases where imports cascade. Let A, B, C be modules where A imports func B.f and rexports A.g, and C import A.g as C.h. When h is called, we need to recursively follow until a local func is hit (which we might not). we might need to check the func_idx is not imported itself here, and if it is, follow the other module without a recursive func call (stackless interpreter rule). We should also throw error if the (module_idx,func_idx) ends up the same as local_module_idx, local_func_idx, since that indicates a cyclic reference. But we can also make do with this too and leave it as an issue.